zk/stages/stage_batches_test.go

```diff
@@ -120,3 +96,132 @@ func TestUnwindBatches(t *testing.T) {
 		require.Equal(t, bucketSized[bucket], size, "butcket %s is not empty", bucket)
 	}
 }
+
+func TestfindCommonAncestor(t *testing.T) {
+	blocksCount := 40
+	l2Blocks := createTestL2Blocks(t, blocksCount)
+
+	testCases := []struct {
+		name                  string
+		dbBlocksCount         int
+		dsBlocksCount         int
+		latestBlockNum        uint64
+		divergentBlockHistory bool
+		expectedBlockNum      uint64
+		expectedHash          common.Hash
+		expectedError         error
+	}{
+		{
+			name:             "Successful search (db lagging behind the data stream)",
+			dbBlocksCount:    5,
+			dsBlocksCount:    10,
+			latestBlockNum:   5,
+			expectedBlockNum: 5,
+			expectedHash:     common.HexToHash("0x5"),
+			expectedError:    nil,
+		},
+		{
+			name:             "Successful search (db leading the data stream)",
+			dbBlocksCount:    20,
+			dsBlocksCount:    10,
+			latestBlockNum:   10,
+			expectedBlockNum: 10,
+			expectedHash:     common.HexToHash("0xa"),
+			expectedError:    nil,
+		},
+		{
+			name:           "Failed to find common ancestor block (latest block number is 0)",
+			dbBlocksCount:  10,
+			dsBlocksCount:  10,
+			latestBlockNum: 0,
+			expectedError:  ErrFailedToFindCommonAncestor,
+		},
+		{
+			name:                  "Failed to find common ancestor block (different blocks in the data stream and db)",
+			dbBlocksCount:         10,
+			dsBlocksCount:         10,
+			divergentBlockHistory: true,
+			latestBlockNum:        20,
+			expectedError:         fmt.Errorf("failed to fetch block %d from data stream", 0),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// ARRANGE
+			testDb, tx := memdb.NewTestTx(t)
+			err := hermez_db.CreateHermezBuckets(tx)
+			require.NoError(t, err)
+
+			err = db.CreateEriDbBuckets(tx)
+			require.NoError(t, err)
+
+			hermezDb := hermez_db.NewHermezDb(tx)
+
+			dsBlocks := l2Blocks[:tc.dsBlocksCount]
+			dbBlocks := l2Blocks[:tc.dbBlocksCount]
+			if tc.divergentBlockHistory {
+				dbBlocks = l2Blocks[tc.dsBlocksCount : tc.dbBlocksCount+tc.dsBlocksCount]
+			}
+
+			dsClient := NewTestDatastreamClient(dsBlocks, nil)
+			for _, l2Block := range dbBlocks {
+				require.NoError(t, hermezDb.WriteBlockBatch(l2Block.L2BlockNumber, l2Block.BatchNumber))
+				require.NoError(t, rawdb.WriteCanonicalHash(tx, l2Block.L2Blockhash, l2Block.L2BlockNumber))
+			}
+			require.NoError(t, tx.Commit())
+
+			erigonDbReader := NewTestErigonDbReader(context.Background(), testDb)
+			hermezDbReader := NewTestHermezDbReaderWrapper(context.Background(), hermez_db.NewHermezDbReader(tx), testDb)
```

Comment:
I've created this wrappers, just because in order to query the MDBX I need to wrap it with  `db.View(ctx, func(tx ...) {...})`. However I would be keen to change it if there is a more elegant approach there.

cc @hexoscott @cffls

Reply:
Hi @Stefan-Ethernal, there are a few things that can help here and some extras:
- The test function name was picked up by the test runner for me until I renamed it to `Test_` with the underscore, or capital `F`.
- When you create the `testDb` and `tx` you should defer the close of the db and the rollback on the tx.
- The commit is the problem here as it closes the transaction down.  If you remove the commit on the transaction you will find the test can just use the `hermezDb.HermezDbReader` without needing the wrapper, and you can use the normal `eridb := erigon_db.NewErigonDb(tx)` for the other case to remove the need for the wrapper.

Reply:
Thanks @hexoscott!

> The test function name was picked up by the test runner for me until I renamed it to Test_ with the underscore, or capital F.

I've named it that way, because of this dull Sonarcloud error (https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=886&id=0xPolygonHermez_cdk-erigon&open=AZEnnvAFgHp7PU-smpPO), but if you agree, let's ignore it.

Reply:
That seems to be a pretty normal way to name test functions to me so happy to ignore the sonar cloud if we can.  If not then just uppercasing the F should make it happy.  Glad the ideas above helped :D